experiments/storm/src/main/scala/io/gearpump/experiments/storm/util/StormOutputCollector.scala

```diff
@@ -18,34 +18,78 @@
   private val LOG: Logger = LogUtil.getLogger(classOf[StormOutputCollector])
+  private val EMPTY_LIST: JList[Integer] = new JArrayList[Integer](0)
 }
 
-class StormOutputCollector(context: TaskContext, pid: Int, componentId: String) {
-  import StormOutputCollector._
+class StormOutputCollector(
+    context: TaskContext,
+    topologyContext: TopologyContext) {
+  import io.gearpump.experiments.storm.util.StormOutputCollector._
 
+  private val stormTaskId = topologyContext.getThisTaskId
+  private val componentId = topologyContext.getThisComponentId
   private var timestamp: TimeStamp = LatestTime
+  private val taskToComponent = topologyContext.getTaskToComponent
+  private val taskToPartition = taskToComponent.map { case (task, component) =>
+    task -> topologyContext.getComponentTasks(component).indexOf(task)
+  }.toMap
+  private val targets = topologyContext.getTargets(componentId)
 
   def emit(streamId: String, values: JList[AnyRef]): JList[Integer] = {
-    context.output(Message(StormTuple(values.toList, pid, componentId, streamId), timestamp))
-    EMPTY_LIST
+    if (targets.containsKey(streamId)) {
+      val outFields = topologyContext.getComponentOutputFields(componentId, streamId)
+      val globalStreamId = new GlobalStreamId(componentId, streamId)
+      val sourceTasks = topologyContext.getComponentTasks(componentId)
+      val targetPartitions = targets.get(streamId).foldLeft(Map.empty[String, List[Int]]) { case (accum, iter) =>
+        val (target, grouping) = iter
+          val groupingContext = GroupingContext(topologyContext, globalStreamId, sourceTasks, outFields, taskToPartition)
+          val targetTasksNum = topologyContext.getComponentTasks(target).size
+          val grouper = getGrouper(groupingContext, grouping, targetTasksNum)
+          val partitions = grouper.getPartitions(stormTaskId, values)
+          accum + (target -> partitions)
+      }
+      // convert backtype.storm.tuple.Values to scala List to avoid serialization issues
+      // TODO: investigate performance impact
+      val tuple = StormTuple(values.toList, componentId, streamId, stormTaskId, targetPartitions)
+      context.output(Message(tuple, timestamp))
+      targetPartitions.toList.flatMap { case (target, partitions) =>
+        partitions.map(topologyContext.getComponentTasks(target).get)
+      }
+    } else {
+      EMPTY_LIST
+    }
   }
 
-  def emitDirect(taskId: Int, streamId: String, tuple: JList[AnyRef]): Unit = {
-    LOG.error("emit direct not supported")
+  def emitDirect(id: Int, streamId: String, values: JList[AnyRef]): Unit = {
+    if (targets.containsKey(streamId)) {
+      val target = taskToComponent(id)
+      val partition = taskToPartition(id)
+      val targetPartitions = Map(target -> List(partition))
+      val tuple = StormTuple(values.toList, componentId, streamId, stormTaskId, targetPartitions)
+      context.output(Message(tuple, timestamp))
+    }
   }
 
+  def fail(tuple: Tuple): Unit = {}
```

Comment:
Can you comment why here it is empty?
Is this expected? or a TODO? 

Will there impacts?